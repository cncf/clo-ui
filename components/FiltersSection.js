module.exports = require('../dist/cjs/components/FiltersSection.js');
