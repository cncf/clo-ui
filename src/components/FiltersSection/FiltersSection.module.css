.categoryTitle {
  margin-top: 2rem;
}

.decorator {
  top: 3px;
}

label {
  font-size: 0.9rem;
}

.search {
  box-shadow: 0 0 0 0.3rem var(--clo-secondary-15) !important;
  margin: 0rem !important;
}

.searchWithClose input {
  max-width: calc(100% - 70px);
}

.subtitle {
  font-size: 0.75rem;
}
