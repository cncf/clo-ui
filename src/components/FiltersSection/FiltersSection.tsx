import classnames from 'classnames';
import { isArray, isNull, isUndefined } from 'lodash';
import React, { ChangeEvent, forwardRef, Ref, useEffect, useImperativeHandle, useState } from 'react';

import { CheckBox } from '../Checkbox/Checkbox';
import { Searchbar } from '../Searchbar/Searchbar';
import styles from './FiltersSection.module.css';

export interface FilterSection {
  key?: string;
  placeholder?: string;
  title: string;
  options:
    | FilterOption[]
    | {
        [key: string]: FilterOption[];
      };
}

export interface FilterOption {
  name: string;
  value?: string;
  type?: string;
  key?: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface RefFiltersSection {
  cleanValue: () => void;
}

export interface IFiltersSectionProps {
  section: FilterSection;
  visibleTitle: boolean;
  contentClassName?: string;
  device: string;
  activeFilters?: string[];
  withSearchBar?: boolean;
  disabled?: boolean;
  onChange: (name: string, value: string, checked: boolean, type?: string) => void;
}

const SEARCH_DELAY = 3 * 100; // 300ms

export const FiltersSection = forwardRef<RefFiltersSection, IFiltersSectionProps>(
  (props: IFiltersSectionProps, ref: Ref<RefFiltersSection>) => {
    const [value, setValue] = useState<string>('');
    const [visibleOptions, setVisibleOptions] = useState<FilterOption[] | { [key: string]: FilterOption[] }>(
      props.section.options
    );
    const [searchTimeout, setSearchTimeout] = useState<NodeJS.Timeout | null>(null);

    useImperativeHandle(ref, () => ({
      cleanValue: () => {
        setValue('');
      },
    }));

    const filterOptions = () => {
      if (props.section.options) {
        if (value !== '') {
          if (isArray(props.section.options)) {
            setVisibleOptions(
              props.section.options.filter(
                (f: FilterOption) =>
                  (f.value && f.value.toLowerCase().includes(value.toLowerCase())) ||
                  f.name.toLowerCase().includes(value.toLowerCase())
              )
            );
          } else {
            const options: { [key: string]: FilterOption[] } = {};
            Object.keys(props.section.options).forEach((group: string) => {
              const optsToCkeck = (props.section.options as { [key: string]: FilterOption[] })[group];

              const opts = optsToCkeck.filter(
                (f: FilterOption) =>
                  (f.value && f.value.toLowerCase().includes(value.toLowerCase())) ||
                  f.name.toLowerCase().includes(value.toLowerCase())
              );
              if (opts.length > 0) {
                options[group] = opts;
              }
            });
            setVisibleOptions(options);
          }
        } else {
          setVisibleOptions(props.section.options);
        }
      }
    };

    useEffect(() => {
      setVisibleOptions(props.section.options);
    }, [props.section.options]);

    useEffect(() => {
      if (!isNull(searchTimeout)) {
        clearTimeout(searchTimeout);
      }
      setSearchTimeout(
        setTimeout(() => {
          filterOptions();
        }, SEARCH_DELAY)
      );
    }, [value]);

    useEffect(() => {
      return () => {
        if (searchTimeout) {
          clearTimeout(searchTimeout);
        }
      };
    }, []);

    return (
      <>
        {props.visibleTitle && (
          <div className={`fw-bold text-uppercase text-primary ${styles.categoryTitle}`}>
            <small>{props.section.title}</small>
          </div>
        )}

        {!isUndefined(props.withSearchBar) && props.withSearchBar && (
          <div className="mb-3">
            <Searchbar
              value={value}
              onValueChange={(newValue: string) => setValue(newValue)}
              onSearch={filterOptions}
              cleanSearchValue={() => setValue('')}
              classNameSearch={classnames(styles.search, { [styles.searchWithClose]: value !== '' })}
              placeholder={`Search ${props.section.key || ''}`}
              bigSize={false}
              disabled={props.disabled}
            />
          </div>
        )}

        <div className={classnames(props.contentClassName, { 'mt-2': props.visibleTitle })}>
          {isArray(visibleOptions) ? (
            <>
              {visibleOptions.map((filter: FilterOption) => {
                if (isNull(filter.name)) return null;
                return (
                  <CheckBox
                    key={`filter_${filter.key || filter.value}`}
                    name={(filter.key || props.section.key)!}
                    value={(filter.key || filter.value)!}
                    labelClassName="mw-100"
                    legend={filter.legend}
                    label={filter.name}
                    icon={<span className={`position-relative ${styles.decorator}`}>{filter.decorator}</span>}
                    device={props.device}
                    checked={
                      !isUndefined(props.activeFilters) && props.activeFilters.includes((filter.key || filter.value)!)
                    }
                    onChange={(e: ChangeEvent<HTMLInputElement>) =>
                      props.onChange(e.target.name, e.target.value, e.target.checked, filter.type)
                    }
                    disabled={props.disabled}
                  />
                );
              })}
            </>
          ) : (
            <>
              {Object.keys(visibleOptions).map((group: string) => {
                if (visibleOptions[group].length === 0) return null;
                return (
                  <React.Fragment key={group}>
                    <div className={`text-uppercase text-muted fw-bold mb-2 mt-4 ${styles.subtitle}`}>{group}</div>
                    {visibleOptions[group].map((filter: FilterOption) => {
                      if (isNull(filter.name)) return null;
                      return (
                        <CheckBox
                          key={`filter_${filter.key || filter.value}`}
                          name={(filter.key || props.section.key)!}
                          value={(filter.key || filter.value)!}
                          labelClassName="mw-100"
                          legend={filter.legend}
                          label={filter.name}
                          icon={<span className={`position-relative ${styles.decorator}`}>{filter.decorator}</span>}
                          device={props.device}
                          checked={
                            !isUndefined(props.activeFilters) &&
                            props.activeFilters.includes((filter.key || filter.value)!)
                          }
                          onChange={(e: ChangeEvent<HTMLInputElement>) =>
                            props.onChange(e.target.name, e.target.value, e.target.checked, filter.type)
                          }
                          disabled={props.disabled}
                        />
                      );
                    })}
                  </React.Fragment>
                );
              })}
            </>
          )}
        </div>
      </>
    );
  }
);
