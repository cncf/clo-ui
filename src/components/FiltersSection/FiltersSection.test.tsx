import { render, screen } from '@testing-library/react';

import { FiltersSection } from './FiltersSection';

const onChangeMock = jest.fn();

const defaultProps = {
  section: {
    key: 'foundation',
    title: 'Foundation',
    options: [
      { value: 'cncf', name: 'CNCF' },
      { value: 'lfaidata', name: 'LF AI & Data' },
    ],
  },
  visibleTitle: false,
  device: 'desktop',
  activeFilters: [],
  onChange: onChangeMock,
};

const openMock = jest.fn();
window.open = openMock;

describe('FiltersSection', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<FiltersSection {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(<FiltersSection {...defaultProps} />);

    expect(screen.getAllByRole('checkbox')).toHaveLength(2);
  });
});
