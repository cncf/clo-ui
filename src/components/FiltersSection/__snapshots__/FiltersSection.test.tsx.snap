// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FiltersSection creates snapshot 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 checkbox"
        id="desktop-foundation-cncf"
        name="foundation"
        tabindex="0"
        type="checkbox"
        value="cncf"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="desktop-foundation-cncf"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative decorator"
          />
          <span
            class="text-truncate"
          >
            CNCF
          </span>
        </div>
      </label>
    </div>
    <div
      class="form-check me-sm-2 mb-2 undefined"
    >
      <input
        aria-checked="false"
        class="form-check-input rounded-0 checkbox"
        id="desktop-foundation-lfaidata"
        name="foundation"
        tabindex="0"
        type="checkbox"
        value="lfaidata"
      />
      <label
        class="form-check-label mw-100"
        data-testid="checkboxLabel"
        for="desktop-foundation-lfaidata"
      >
        <div
          class="d-flex align-items-baseline mw-100"
        >
          <span
            class="position-relative decorator"
          />
          <span
            class="text-truncate"
          >
            LF AI & Data
          </span>
        </div>
      </label>
    </div>
  </div>
</DocumentFragment>
`;
